### DISCUSSION TITLE
Unable to sync after login to PopOS -- database locked by another process

### DISCUSSION THREAD

**Discussion by salaivv** [abraunegg/onedrive#1272] [Q&A] [answered]
Hi @abraunegg ,

I recently posted a discussion about migrating to linux from windows. Unfortunately I am not able to remove Windows since I have a couple apps only available on windows. 

Nonetheless I added a separate SSD to my system and installed PopOS on it. And since I am currently on a dual boot sharing data between Windows and Linux, my only choice is to use NTFS for the data partition. (couldn't afford to make the switch to ext4 or ZFS right now :) )

Now the problem is, whenever I login to PopOS, I could see that the `onedrive --monitor` process is writing around 1.3 GB of data to my disk. This happens at every startup and during that time if I check sync status using `onedrive --display-sync-status`, it shows that my local directory is out of sync and that approximately a gigabyte of data needs to be uploaded (this size depends on how soon I check sync status after I login). If I try to synchronize or re-synchronize, I get an error that the database is locked or I get a disk I/O error. 

But once the monitor stops writing data to disk (stops around 1.3-1.4 GB), then sync works normally and reports that the local files are in sync and nothing is pending.

1. I kind of understand that the `onedrive --monitor` process is the one that has a lock on the database. But I wonder if this behavior is normal. Or is this a problem with using NTFS on linux?

2. Also I never used `--monitor` switch to sync. But onedrive automatically starts the sync engine and the monitor on login. Is this normal behavior as well? I expected onedrive to run only when I run the sync command on the terminal. Or am I missing something here?

*A small note -- I installed onedrive from the Ubuntu PPA mentioned in the INSTALL.md document.*

*And my data partition is on the hard drive in which Windows is installed. There are two partitions -- one for windows and one for data where my sync folder is located and PopOS is on a separate SSD.*

**Comment by abraunegg** [comment, marked as answer]
@salaivv 
Something is *not* right here.

There are *known* problems / issues with the Ubuntu PPA package with it automatically starting, but also unsure what else the package is doing. I am not the package maintainer, so I cannot look at or fix any issue with the package.

The client however will (regardless), at startup, perform an integrity check of data - but, if the client is downloading files at every startup, then this is potentially a consequence of also using NTFS for your sync_dir as PopOS might not be able to read|write the files correctly, thus, the local client file cache thinks things are out of sync. The client might not be able to update / write out the data to cache database either - this is stored in your client configuration directory.

> If I try to synchronize or re-synchronize, I get an error that the database is locked or I get a disk I/O error.

This is because there is already a process running performing the sync. Dont use `--resync` unless the application advises you to. To avoid this message you have to stop the systemd process that is already running, but if you did not start it - how do you go and stop it ... PPA package issue.

A couple of suggestions:
1. > I recently posted a discussion about migrating to linux from windows. Unfortunately I am not able to remove Windows since I have a couple apps only available on windows.

    Use VMware Workstation | Virtualbox | KVM as a application under PopOS, to run your Windows application and run your Windows only applications that way. If it is 'gaming' applications, VMware Workstation will probably be best given the graphics acceleration, if something else, then any of these will do.

2. File a bug against the Ubuntu PPA package

    The package should *not* be auto starting / auto configuring. It should just install the files and allow you, the user, to configure / setup systemd services once your configuration and testing is complete. File a bug here regarding the issue, and file a bug [here](https://launchpad.net/~yann1ck/+archive/ubuntu/onedrive)

3. Uninstall the Ubuntu PPA package and build from source.

    This will let you control when and how the service is starting, and ensure that the right options are being used at startup. Essentially what you should be doing is making sure your client works 100% the way you want with `--synchronize --verbose --dry-run` before enabling *any* systemd service. 

    If going down this route, you *will* also need to ensure you perform a *full* package cleanup to ensure no unwarranted file is being left behind. If you are also still working out what distribution to use (hence the separate SSD), I actually would suggest pick something that is not Ubuntu based. 

    If you want to use the `apt` package / process etc, I suggest you look at Debian (Ubuntu is based off Debian, but Ubuntu change things), or if you want to use something more commercial grade - RedHat changed their RHEL licencing to no-cost (free) for small production workloads (up to 16 systems) - see [this](https://www.redhat.com/en/blog/new-year-new-red-hat-enterprise-linux-programs-easier-ways-access-rhel) blog post. On a side note, I have been using RedHat since ~1996 and in various forms ever since.

**Reply by salaivv** [reply]
Hi @abraunegg, I removed the PPA package and built onedrive from source. And like you mentioned onedrive is working as expected and doesn't start automatically at startup. :+1: 

> The client however will (regardless), at startup, perform an integrity check of data.

Is there any way to know this is happening? I couldn't see any notification or looking at the system monitor, no onedrive process seems to be running in the background. But nonetheless, sync is working perfectly as expected.

Also, now that I built onedrive from source, when there is a new release can I just pull changes, build again and do `make install` and will it upgrade automatically? Or shall I have to uninstall first before upgrading?

> File a bug here regarding the issue, and file a bug here.

You mean both here in your github repository and the PPA?

> If you are also still working out what distribution to use (hence the separate SSD), I actually would suggest pick something that is not Ubuntu based.

The only reason I picked up PopOS is that it supports NVIDIA cards out of the box since it comes with the drivers preinstalled. Most of my work is on Blender / Unreal Engine and these apps depend on GPU a lot.

> if you want to use something more commercial grade - RedHat changed their RHEL licencing to no-cost (free) for small production workloads (up to 16 systems) - see this blog post.

But I definitely would like to run a RHEL based distro. I am actually waiting for the Rocky Linux release. I am just reluctant about registering or managing subscriptions. :)

It should be fine, right? Or will it also have a few things changed like you mentioned for the Debian -- Ubuntu distros. May be not the right time to speculate about Rocky since we don't even have an RC yet. But may be anything about CentOS for that matter?

And, once again, thank you so much for helping out. :)

**Reply by abraunegg** [reply]
@salaivv 

> And like you mentioned onedrive is working as expected and doesn't start automatically at startup. 👍
> 
> > The client however will (regardless), at startup, perform an integrity check of data.

Let me clarify this: the onedrive client, when configured as a systemd service will (regardless), at startup of the client application, perform an integrity check of data.

I was not meaning at system startup.

> Also, now that I built onedrive from source, when there is a new release can I just pull changes, build again and do make install and will it upgrade automatically? Or shall I have to uninstall first before upgrading?

You will have to watch the repo for when new releases are done, then rebuild and re-install your client. No need to uninstall. The alternate here is to get the Ubuntu PPA fixed, or, become the Ubuntu package maintainer and fix it properly.

> You mean both here in your github repository and the PPA?

Yes here and against the PPA .. this way, the packaging issue has visibility, and hopefully gets fixed.

> The only reason I picked up PopOS is that it supports NVIDIA cards out of the box since it comes with the drivers preinstalled. Most of my work is on Blender / Unreal Engine and these apps depend on GPU a lot.

Whilst RHEL does not include it by default, this solution document details how to do it: https://access.redhat.com/solutions/4134381

> But I definitely would like to run a RHEL based distro. I am actually waiting for the Rocky Linux release. I am just reluctant about registering or managing subscriptions. :)
>
> It should be fine, right?

It should be fine - a developer account, using an email account specific for just that purpose - so if that account is leaked / compromised or whatever, downstream impacts are minimal - though if RedHat have an exposure like that, something bigger to be worried about.

**Reply by salaivv** [reply]
Thank you so much, @abraunegg.

> Yes here and against the PPA .. this way, the packaging issue has visibility, and hopefully gets fixed.

Sure! I'll file a bug here and in the PPA as well.

> Whilst RHEL does not include it by default, this solution document details how to do it: https://access.redhat.com/solutions/4134381

And thank you so much for this link. I am looking into it.
